### PR TITLE
chore: use PIPE redirect strategy and consume stream when running external processes

### LINKED ISSUE #13655: Vaadin 23 Dev Mode + Payara -> NPM Dependency Error
### Description of the bug

Deploying a simple Vaadin application in dev mode to Payara Server 5.2022.2 causes NPM Dependency ERROR. 
Deploying the same application to Tomcat works fine.
Deploying the same application in production mode works with both servers.

### Expected behavior

Dev Mode should be available with Vaadin 23 and Payara.

### Minimal reproducible example

Preparation
1. generate default app start.vaadin.com 
2. publish to github (https://github.com/skiedrowski/vaadinDevModeTrouble)
3. modify for war deployment (https://vaadin.com/blog/packaging-and-deploying-spring-boot-applications-as-war-files)
4. push to github

Tomcat works
1. clone repo https://github.com/skiedrowski/vaadinDevModeTrouble into `vaadinDevModeTroubleTomcat`
2. run `./mvnw package`
3. Download and extract Tomcat 9.0.62 into project
4. start Tomcat (`cd apache-tomcat-9.0.62/bin/ && ./startup.sh`)
5. Deploy to Tomcat (`cd ../.. && cp target/myapp.war apache-tomcat-9.0.62/webapps/`)
6. open http://localhost:8080/myapp/
--> works fine
7. shutdown `apache-tomcat-9.0.62/bin/shutdown.sh`

Payara fails
1. clone repo https://github.com/skiedrowski/vaadinDevModeTrouble into `vaadinDevModeTroublePayara`
2. run `./mvnw package`
3. Download and extract Payara 5.2022.2 into project
4. start Payara (`payara5/bin/asadmin start-domain`)
5. Deploy to Payara (`payara5/bin/asadmin deploy target/myapp.war`)
6. open http://localhost:8080/myapp/

> Whitelabel Error Page
>
> This application has no explicit mapping for /error, so you are seeing this as a fallback.
> 
> Fri Apr 29 08:01:53 CEST 2022
> There was an unexpected error (type=Internal Server Error, status=500).

[payara5/glassfish/domains/domain1/logs/server.log](https://github.com/vaadin/flow/files/8589063/server.log) contains
```
[2022-04-29T08:49:45.257+0200] [Payara 5.2022.2] [INFORMATION] [] [] [tid: _ThreadID=101 _ThreadName=ForkJoinPool.commonPool-worker-19] [timeMillis: 1651214985257] [levelValue: 800] [[
  2022-04-29 08:49:45.257 ERROR 48064 --- [nPool-worker-19] c.v.f.s.frontend.TaskUpdatePackages      : Command `/Users/sascha/.vaadin/node/node /Users/sascha/.vaadin/node/node_modules/npm/bin/npm-cli.js --no-update-notifier --no-audit --scripts-prepend-node-path=true --ignore-scripts install` failed:

]]

[2022-04-29T08:49:45.257+0200] [Payara 5.2022.2] [INFORMATION] [] [] [tid: _ThreadID=101 _ThreadName=ForkJoinPool.commonPool-worker-19] [timeMillis: 1651214985257] [levelValue: 800] [[
  2022-04-29 08:49:45.257 ERROR 48064 --- [nPool-worker-19] c.v.f.s.frontend.TaskUpdatePackages      : >>> Dependency ERROR. Check that all required dependencies are deployed in npm repositories.
]]

[2022-04-29T08:49:55.254+0200] [Payara 5.2022.2] [INFORMATION] [] [] [tid: _ThreadID=112 _ThreadName=http-thread-pool::http-listener-1(5)] [timeMillis: 1651214995254] [levelValue: 800] [[
  2022-04-29 08:49:55.254 ERROR 48064 --- [p-listener-1(5)] c.v.flow.server.DefaultErrorHandler      : 

java.lang.IllegalStateException: com.vaadin.flow.server.ExecutionFailedException: Npm install has exited with non zero status. Some dependencies are not installed. Check npm command output
	at com.vaadin.base.devserver.AbstractDevServerRunner.getCause(AbstractDevServerRunner.java:769) ~[vaadin-dev-server-23.0.5.jar:na]
	at com.vaadin.base.devserver.AbstractDevServerRunner.getCause(AbstractDevServerRunner.java:765) ~[vaadin-dev-server-23.0.5.jar:na]
	at com.vaadin.base.devserver.AbstractDevServerRunner.handleRequest(AbstractDevServerRunner.java:631) ~[vaadin-dev-server-23.0.5.jar:na]
	at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1567) ~[flow-server-23.0.5.jar:23.0.5]
	at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:299) ~[flow-server-23.0.5.jar:23.0.5]
...
Caused by: com.vaadin.flow.server.ExecutionFailedException: Npm install has exited with non zero status. Some dependencies are not installed. Check npm command output
...
```
[~/.npm/_logs/2022-04-29T06_49_42_618Z-debug-0.log](https://github.com/vaadin/flow/files/8589059/2022-04-29T06_49_42_618Z-debug-0.log) contains
```
272 warn deprecated source-map-url@0.4.1: See https://github.com/lydell/source-map-url#deprecated
273 timing reifyNode:node_modules/source-map-url Completed in 1755ms
274 verbose stack Error: write EPIPE
274 verbose stack     at afterWriteDispatched (node:internal/stream_base_commons:160:15)
274 verbose stack     at writeGeneric (node:internal/stream_base_commons:151:3)
274 verbose stack     at Socket._writeGeneric (node:net:795:11)
274 verbose stack     at Socket._write (node:net:807:8)
274 verbose stack     at writeOrBuffer (node:internal/streams/writable:389:12)
274 verbose stack     at _write (node:internal/streams/writable:330:10)
274 verbose stack     at Socket.Writable.write (node:internal/streams/writable:334:10)
274 verbose stack     at EventEmitter.log.write (/Users/sascha/.vaadin/node/node_modules/npm/node_modules/npmlog/lib/log.js:362:10)
274 verbose stack     at EventEmitter.<anonymous> (/Users/sascha/.vaadin/node/node_modules/npm/node_modules/npmlog/lib/log.js:297:12)
274 verbose stack     at Array.forEach (<anonymous>)
275 verbose cwd /Users/sascha/zitecs/src/com.zitecs/ologis/core/vaadinDevModeTroublePayara
276 verbose Darwin 21.4.0
277 verbose argv "/Users/sascha/.vaadin/node/node" "/Users/sascha/.vaadin/node/node_modules/npm/bin/npm-cli.js" "--no-update-notifier" "--no-audit" "--scripts-prepend-node-path=true" "--ignore-scripts" "install"
278 verbose node v16.14.2
279 verbose npm  v8.5.0
280 error code EPIPE
281 error syscall write
282 error errno -32
283 error write EPIPE
284 verbose exit -32
```

7. shutdown `payara5/bin/asadmin stop-domain`

### Versions

- Vaadin / Flow version: reproduced with 23.0.5 and 23.0.7
- Java version: 11.0.14, 11.0.11
- OS version: macOS 12.3.1, macOS 10.15.5
- Hardware: MacBook Pro M1 Max, MacBook Pro Intel i9
- Browser version: Safari 15.4 and Chrome 101.0.4951.41
- Application Server: reproduced with Payara Server (Full) 5.2022.2 and Payara Server (Full) 5.2020.6
- IDE (if applicable): not relevant (IntelliJ IDEA 2022.1 Ultimate)

### PR DESCRIPTION
**Pull request by mcollovati**
When running an external process using INHERIT redirect strategy may break in some situations as the output is written directly to the main process stream.
Errors may also arise on some OS when the strategy is PIPE (the default) and the streams are not completely consumed, as the call for `Process.waitFor` may hang indefinitely.
This change uses PIPE for external processes and consumes both input and error streams of the started process.

Fixes #13655 and #15339

### CODE REVIEW

File: flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendToolsLocator.java [hunk 1]
```diff
@@ -158,34 +165,41 @@ private Optional<CommandResult> executeCommand(boolean logErrorOnFail,
...
         try {
-            String stream = IOUtils.toString(process.getInputStream(),
-                    StandardCharsets.UTF_8);
-            if (!stream.isEmpty()) {
-                Stream.of(stream.split("\\R")).forEach(stdout::add);
+            Pair<String, String> outputs = streamConsumer.get();
+            stdout = new ArrayList<>(
+                    List.of(outputs.getFirst().split(System.lineSeparator())));
+            stderr = new ArrayList<>(
+                    List.of(outputs.getFirst().split(System.lineSeparator())));
```
Comment (czp13) [inline flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendToolsLocator.java:176]: I think here we should look for the second pair:
```suggestion
            stderr = new ArrayList<>(
                    List.of(outputs.getSecond().split(System.lineSeparator())));
```

Checking the implementation of the consumeProcessStreams method in FrontendUtils:
```Java
    public static CompletableFuture<Pair<String, String>> consumeProcessStreams(
            Process process) {
     ...
        return CompletableFuture.allOf(stdOut, stdErr).thenApply(
                unused -> new Pair<>(stdOut.getNow(""), stdErr.getNow("")));
    }
```
It seems the second element of the Pair is the error we are looking for here.

Comment (mcollovati) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendToolsLocator.java:176]: Good catch!

File: flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendToolsLocator.java [hunk 2]
```diff
@@ -158,34 +165,41 @@ private Optional<CommandResult> executeCommand(boolean logErrorOnFail,
...
+                    commandString, cause);
             return Optional.empty();
         }
 
-        List<String> stderr = new ArrayList<>();
+        return Optional.of(new CommandResult(commandString, process.exitValue(),
+                stdout, stderr));
+    }
+
+    private List<String> collectOutput(InputStream inputStream) {
```
Comment (czp13) [inline flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendToolsLocator.java:191]: The method is not used. Do we want to add this back somewhere in the `executeCommand()` method logic in this part? Or can we safely delete this?

Comment (czp13) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendToolsLocator.java:191]: I guess some part could be used like `out.addAll(Arrays.asList(stream.split("\\R")));`

**Review by czp13 — COMMENTED** [review]
I think it is a good improvement, and shall fix the referenced issue! 👍  
Although I had some questions, and confusion from my side, and would be good to clean those up before approving it.

Also, could we, should we add covering tests here? I would assume no test coverage for these methods.
(not necessary for approval, just a question :))

File: flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendToolsLocator.java (hunk 2 above)
Comment (mcollovati) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendToolsLocator.java:191]: The method could be removed

**Comment by czp13** [comment]
Wow. Thanks for tackling the ideas, also kudos for adding the tests!
Seems hard to come up with these tests, and look super nice and indeed cover many cases! 🙇

**Review by czp13 — APPROVED** [review]

**Comment by mcollovati** [comment]
The "slow tests" passed on the PR. So, I wonder why it is blocking after the change was merged

**Comment by czp13** [comment]
Is it possible that slow tests became flaky? As in some PR-s the slow test are passing like this PR:
- https://github.com/vaadin/flow/pull/16667

**Pull request merged**